### PR TITLE
Refactor Settings pages

### PR DESCRIPTION
**Pull request by pbiggar**
I went to add a settings page to make it easier to use a tunnel for local assets, and settings was a mess so this refactors it.

I started to add a contributor page but I disabled it for now, as it's not finished, and I'd like to get the refactor out first.

We now have a situation where settings is it's own component, and the three sub-tabs are their own components as well, which know nothing about the components into which they are embedded.

I initially tried to split each component into three files: view, update, and state. The view split was necessary because views use types from AppTypes. I tried to fix this but it turns out it was pretty hard, due to events having `AppTypes.cmd` types, which I struggled to make generic. I managed to re-combine the state and update files, which made things simpler, but I left views as they were.

I added a new concept concept called Effects, to allow components to send back future work to the event loop, which they are not able to do directly, because of dependency ordering. I've written a doc explaining this, included, so I let that stand for itself rather than re-explaining it. I added CCC (Cross-Component Calls) which are intended to replaced modifications as we componentize more stuff. I'm not sure I have the boundaries exactly right, but things seem to work.

As a followup, I plan to integrate #4343 and continue to add the tunnel option.

### CODE REVIEW

File: client/src/components/Settings/SettingsView.res [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  }
+}
+
+let tabTitleView = (tab: T.Tab.t): Html.html<msg> => {
+  let tabTitle = (t: T.Tab.t) => {
+    let isSameTab = tab == t
+
+    Html.h3(
+      list{
+        Html.classList(list{("tab-title", true), ("selected", isSameTab)}),
```
Comment (StachuDotNet) [inline client/src/components/Settings/SettingsView.res:48]:
```suggestion
    let isCurrentTab = tab == t

    Html.h3(
      list{
        Html.classList(list{("tab-title", true), ("selected", isCurrentTab)}),
```

minor; not strongly held

File: client/src/app/Main.res [hunk 2]
```diff
@@ -213,7 +214,7 @@ let processFocus = (m: model, focus: AppTypes.Focus.t): modification =>
       })
     | (_, _) =>
       switch page {
-      | SettingsModal(tab) => Many(SettingsView.getModifications(m, OpenSettingsView(tab)))
+      // | SettingsModal(tab) => Many(SettingsUpdate.getModifications(m, Open(tab)))
       | _ => NoChange
```
Comment (StachuDotNet) [inline client/src/app/Main.res:218]: I think this can be simplified to just `NoChange`?

Comment (StachuDotNet) [reply client/src/app/Main.res:218]: Ah, actually - this is likely relevant to directing to the modal when the page opens. If so, ignore.

**Review by StachuDotNet — COMMENTED** [review]
Looks great; just a few notes, will review again once the mentioned follow-ups are in. :+1:

File: client/src/components/Settings/SettingsView.res (hunk 1 above)
Comment (pbiggar) [reply client/src/components/Settings/SettingsView.res:48]: These were already like this

File: client/src/app/Main.res (hunk 2 above)
Comment (pbiggar) [reply client/src/app/Main.res]: I pushed some new commits to fix this. This is where the settings modal is loaded if you go to a `/settings/` url.

**Review by StachuDotNet — APPROVED** [review]

**Pull request merged**